### PR TITLE
Add Job.status.uncountedTerminatedPods for Job tracking

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

/sig apps
/area batch
/area workload-api/job

Adds Job.status.uncountedTerminatedPods for tracking Job Pods that have finished but are not yet counted as failed or succeeded

And feature gate JobTrackingWithFinalizers

**Which issue(s) this PR fixes**:

Ref kubernetes/enhancements#2307 #28486

Implementation PR to follow.

**Does this PR introduce a user-facing change?**:

```release-note
Added field .status.uncountedTerminatedPods to the Job resource. This field is used by the job controller to keep track of finished pods before adding them to the Job status counters.

Pods created by the job controller get the finalizer batch.kubernetes.io/job-tracking

Jobs that are tracked using this mechanism get the annotation batch.kubernetes.io/job-tracking. This is a temporary measure. Two releases after this feature graduates to beta, the annotation won't be added to Jobs anymore.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: git.k8s.io/enhancements/keps/sig-apps/2307-job-tracking-wihout-lingering-pods
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/job
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
Implementation PR is ready #98817

File: pkg/registry/batch/job/strategy_test.go
Comment (soltysh) [inline pkg/registry/batch/job/strategy_test.go:42]: Nit: let's keep the naming consistent, how about `trackingWithFinalizersEnabled`?

File: pkg/apis/batch/types.go
Comment (soltysh) [inline pkg/apis/batch/types.go:259]: There's a mismatch between this description and what the storage layer does. This says when it's nil the controller doesn't use this feature, which means when the feature was on and job was created and then the feature was off the field should be nil again, right? So you either:
1. change the description to account for the above situation
2. change the storage layer (`pkg/registry/batch/job/strategy.go`) such that in `PrepareForUpdate` it will set the field to nil.

**Review by soltysh — COMMENTED** [review]
A few nits + rebase required.

File: pkg/apis/batch/types.go
Comment (ravisantoshgudimetla) [reply pkg/apis/batch/types.go:261]: The opposite of 2nd scenario could also happen, for example `oldObject.field != nil and featureGate is not enabled` momentarily in the old apiserver which has not been rolledout, we would preserve the value associated with oldObj and controller may act upon. To me, kube is eventually consistent and there will be stages which may not be desired in reconciliation. The entities waiting on the job workloads should wait for certain time before those reconciliation happens or timeout.

 If you think null is a condition we explicitly in other scenarios, then perhaps it is ok to put it in the comments, if not I'd drop it.

**Review by ravisantoshgudimetla — CHANGES_REQUESTED** [review]
I've some questions around the comment but rest LGTM

File: pkg/apis/batch/types.go
Comment (soltysh) [reply pkg/apis/batch/types.go:261]: I like that suggestion from @ravisantoshgudimetla and I think we are using that in other places already.

Comment (soltysh) [reply pkg/apis/batch/types.go:261]: Actually, on a second thought - yeah I think @alculquicondor point is strong enough, but we need to make sure this is properly documented :sweat_smile:

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:261]: Anything you would like to be added to the comment? I still think the description is accurate of what happens.

Comment (ravisantoshgudimetla) [reply pkg/apis/batch/types.go:261]: After having a private conversation with @alculquicondor this LGTM. However the contention point of `overloaded meaning of empty struct in the status` is still valid.

@alculquicondor told that he will think of alternatives during beta stage.

**Review by ravisantoshgudimetla — APPROVED** [review]
/lgtm

Thank you for working on this @alculquicondor

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:261]: The idea is that, if there is a way to start tracking with finalizers an existing job that was previously tracked without them, there is no need for distinguishing nil from empty struct.

If we cannot find a safe way to do that, the distinction can still be removed once we figured that the tracking has existed for long enough that there shouldn't exist jobs tracked with the legacy algorithm.

Comment (alculquicondor) [reply pkg/apis/batch/types.go:261]: @ravisantoshgudimetla proposed one solution to avoid having to distinguish between empty and nil: add the finalizer to the Job through an admission controller.

The updated proposal is in kubernetes/enhancements#2776

**Comment by alculquicondor** [comment]
@ravisantoshgudimetla I implemented kubernetes/enhancements#2776 in the latest commit. PTAL.
The implementation PR is also updated.

**Review by ravisantoshgudimetla — COMMENTED** [review]
LGTM. I just have one question about the field not getting dropped.

File: pkg/registry/batch/job/strategy.go
Comment (alculquicondor) [reply pkg/registry/batch/job/strategy.go]: You commented in an outdated change. Is this still relevant after the last commit?

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by alculquicondor** [comment]
Thanks! Changes are now squashed

**Comment by alculquicondor** [comment]
rebased

File: pkg/registry/batch/job/strategy.go
Comment (roycaihw) [inline pkg/registry/batch/job/strategy.go:108]: the finalizer is used for cleaning up pod finalizers and as a marker (for transition between old and new tracking mechanism), but we discussed that: 
- for cleaning up pod finalizers, the controller that watches pods should be more robust and remove the finalizer if a job is not found
- for the marker, you can use an annotation

using a finalizer in the job may cause more trouble where jobs cannot be deleted.

File: pkg/apis/batch/types.go
Comment (roycaihw) [inline pkg/apis/batch/types.go:260]: the documentation does match with the name of the field

File: pkg/apis/batch/types.go
Comment (roycaihw) [inline pkg/apis/batch/types.go:273]: worth documenting why this is useful

File: pkg/apis/batch/types.go
Comment (liggitt) [inline pkg/apis/batch/types.go:260]:
```suggestion
	// UncountedTerminatedPods holds the UIDs of Pods that have terminated but
```

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:269]:
```suggestion
	// Track Job completion without relying on Pod remaining in the cluster
```

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:270]:
```suggestion
	// indefinitely. Pod finalizers, in addition to a field in the Job
```

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/batch/v1/types.go:28]: rename since this is no longer being used as a finalizer, and add godoc explaining what object this is added to and why

File: pkg/registry/batch/job/strategy.go
Comment (liggitt) [inline pkg/registry/batch/job/strategy.go:134]: can drop the nil check, delete works fine on nil

File: pkg/registry/batch/job/strategy.go
Comment (liggitt) [inline pkg/registry/batch/job/strategy.go:167]: what happens if the annotation gets deleted on update by the client making the API request? does the controller go back to tracking by pod counts? does that work properly (or at least no worse than job completion calculation currently works)? is it tested?

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/batch/v1/types.go:246]:
```suggestion
	// uncountedTerminatedPods holds the UIDs of Pods that have terminated but
```

File: pkg/apis/batch/types.go
Comment (liggitt) [inline pkg/apis/batch/types.go:290]: should these be validated in ValidateJobStatus? ensure no duplicate items, no empty items, etc?

**Review by liggitt — COMMENTED** [review]
looks like openapi needs updating (`Please run hack/update-openapi-spec.sh`)

once this is done, link to the controller PR... we'll want both tested/reviewed/ready before merge

**Review by alculquicondor — COMMENTED** [review]
I hadn't finish yet 😅 I just added JobValidationOptions to handle the annotation

File: pkg/registry/batch/job/strategy.go
Comment (roycaihw) [inline pkg/registry/batch/job/strategy.go:198]: should be `||`?

Comment (alculquicondor) [reply pkg/registry/batch/job/strategy.go:198]: And that is true with this check.

The reason why we require that the old job also has the annotation is that we don't support migrating from a legacy tracking to the new tracking.

**Comment by alculquicondor** [comment]
Btw, the implementation PR only had 2 nits left #98238

File: pkg/registry/batch/job/strategy.go
Comment (roycaihw) [reply pkg/registry/batch/job/strategy.go:198]: IIUC, if I disable the feature, and try to suspend a job which already has the annotation, I will get a "restricted annotation" error, right?

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:27]: Changed to say "until 1.26", projected GA + 1

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:296]: Added

File: pkg/registry/batch/job/strategy.go
Comment (roycaihw) [reply pkg/registry/batch/job/strategy.go:161]: sounds good. Why don't we also return an error when the feature is disabled?

**Review by roycaihw — COMMENTED** [review]
I have one question. LGTM otherwise.

File: pkg/registry/batch/job/strategy.go
Comment (alculquicondor) [reply pkg/registry/batch/job/strategy.go:161]: This is what we do for other gated fields.

I believe the intent is that when the feature was enabled, users might have started to add the field to their spec. We don't want those specs do suddenly start failing.

Comment (alculquicondor) [reply pkg/registry/batch/job/strategy.go:161]: That rationale doesn't apply for a Job that didn't have the the annotation before.

**Comment by roycaihw** [comment]
/lgtm

please squash the commits

File: pkg/apis/batch/types.go
Comment (liggitt) [reply pkg/apis/batch/types.go:296]: duplicate and empty items is mostly to prevent controller accidents that could cause the object to grow without bound

File: pkg/apis/batch/validation/validation.go
Comment (liggitt) [inline pkg/apis/batch/validation/validation.go:92]: would this be a clearer comment?
```suggestion
		allErrs = append(allErrs, field.Forbidden(field.NewPath("metadata").Child("annotations").Key(batch.JobTrackingFinalizer), "cannot add this annotation on update"))
```

**Comment by alculquicondor** [comment]
Squashed

File: pkg/apis/certificates/validation/validation_test.go
Comment (liggitt) [inline pkg/apis/certificates/validation/validation_test.go:193]: surprised this file got touched...

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go:32]: suggest framing this in terms of "X releases after GA of feature Y" instead of a specific release more than a year in the future

I'm also trying to envision the transition when this goes GA and we no longer need to decorate the job to get this behavior... is that sketched out anywhere, at what point the controller will unconditionally track pods this way? How will the controller decide it is safe to switch to using the finalizer pod mechanism on existing jobs that may already have pods created?

**Review by liggitt — COMMENTED** [review]
API changes lgtm.

Just a couple comments on doc strings/messages and one question about the GA transition. Once the implementation PR has lgtm on top of this, both can merge together.

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (alculquicondor) [reply staging/src/k8s.io/api/batch/v1/types.go:32]: Changed to "two releases after GA".

> I'm also trying to envision the transition when this goes GA and we no longer need to decorate the job to get this behavior.

I either have to find a safe way to add the finalizer to existing pods or we wait long enough to make sure that there shouldn't be existing Jobs that are tracked with the legacy behavior (or a combination of both).
Adding finalizers is tricky because we can't tell for sure if a finished pod doesn't have a finalizer because it's from the legacy tracking or because we counted it already. We can probably add the finalizer to running pods.

In our favor, we have that Jobs are run-to-completion. So eventually there shouldn't be non-finished Jobs with legacy behavior. Worst case scenario, we would recreate a few pods, once we get rid of the legacy tracking.

I was deferring these discussions to beta. I will add some notes when I update the KEP to match the implementation (the use of the finalizer as annotation and the extra work queue).

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [reply pkg/apis/batch/validation/validation.go]: I removed the "on update" part, as this function is independent of create/update distinction.

**Comment by alculquicondor** [comment]
FYI the implementation PR is lgtmed and approved https://github.com/kubernetes/kubernetes/pull/98238#pullrequestreview-695276662

I'm rebasing now (which will remove the LGTM)

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (liggitt) [reply staging/src/k8s.io/api/batch/v1/types.go:34]: > I was deferring these discussions to beta. I will add some notes when I update the KEP to match the implementation (the use of the finalizer as annotation and the extra work queue).

a clear picture of what will happen for this at each transition is important to have before we start adding annotations to things and implementing controllers

**Comment by liggitt** [comment]
> I'm also trying to envision the transition when this goes GA and we no longer need to decorate the job to get this behavior... is that sketched out anywhere, at what point the controller will unconditionally track pods this way? How will the controller decide it is safe to switch to using the finalizer pod mechanism on existing jobs that may already have pods created?

Can defer this either to a KEP update or the implementation PR, but you should have a clear picture of the whole lifecycle of this annotation (through to GA and dropping that code path) before you start populating it

**Comment by liggitt** [comment]
API lgtm

**Comment by alculquicondor** [comment]
I want to defer it to a KEP update, given that code freeze is approaching.
Any other thing before approval?

File: staging/src/k8s.io/api/batch/v1/types.go
Comment (soltysh) [reply staging/src/k8s.io/api/batch/v1/types.go:34]: > I either have to find a safe way to add the finalizer to existing pods or we wait long enough to make sure that there shouldn't be existing Jobs that are tracked with the legacy behavior (or a combination of both).

Honestly, that was what I was imagining that we'd keep both ways for more than usual supported time just to give folks the time to complete their jobs. All the new ones will always be tracked with the new mechanism when it gets beta, and old ones will eventually complete, as Aldo mentions. We just need to clearly define how long we'll keep the legacy behavior and explain that in the release notes.

**Comment by alculquicondor** [comment]
Release note updated.

**Comment by alculquicondor** [comment]
ping @liggitt for approval

**Comment by alculquicondor** [comment]
Rebased

**Comment by liggitt** [comment]
/approve
for API bits

**Comment by liggitt** [comment]
integration flake is really bad, tracked in https://github.com/kubernetes/kubernetes/issues/103512

**Comment by alculquicondor** [comment]
Also unit #103549

**Review by soltysh — APPROVED** [review]
/lgtm
/retest

**Comment by alculquicondor** [comment]
Fixed unit test

**Pull request merged**
